### DISCUSSION TITLE
`equals()`  of different numeric `JsonNode` subtypes with same underlying value fails

### DISCUSSION THREAD

**Discussion by Captain-P-Goldfish** [FasterXML/jackson-databind#4343] [Q&A]
### Search before asking

- [X] I searched in the [issues](https://github.com/FasterXML/jackson-databind/issues) and found nothing similar.

### Describe the bug

Hi, I am currently struggling with a little problem when comparing 2 `ObjectNodes`.

![comparisonFailure](https://github.com/FasterXML/jackson-databind/assets/3079540/06576d2e-f29d-43de-ae7c-ef54074eed1d)

The comparison of the two red-surrounded objects does fail. The reason is the `numberArray`.
`ObjectNode`-1 holds these values as `IntNode`
while
`ObjectNode`-2 holds these values as `LongNode`

would it be possible to adjust the comparison to let them match by value instead of type and value?

### Version Information

2.16.0

### Reproduction

see above

### Expected behavior

comparison succeeds

**Comment by JooHyukKim** [comment]
~~I don't think there's such an option nor the comparison methods are overridable. 🤔~~

**Comment by cowtowncoder** [comment]
This is one area where it is quite impractical to try to coerce values to make different number types match, so no attempt is made. This has been extensively discussed in the past.

However, there IS a way to customize this by calling this method on `JsonNode`:

    public boolean equals(Comparator<JsonNode> comparator, JsonNode other);

providing your custom `Comparator<JsonNode>` implementation, which may then use whatever coercion/conversion rules you want. It is meant to allow solving this particular problem.

**Comment by cowtowncoder** [comment]
@JooHyukKim As per above, there is actually a way :)

**Comment by JooHyukKim** [comment]
> @JooHyukKim As per above, there is actually a way :)

Ah, there is [usage example](https://github.com/fasterxml/jackson-databind/issues/790#issuecomment-101509465)

Btw, this seems more like usage question.
Should we transfer this to Dicussions tab, @cowtowncoder ?

**Comment by cowtowncoder** [comment]
Yes, good idea, this is more Discussion than issue.

**Comment by Captain-P-Goldfish** [comment]
Okay, I solved it like this now:
```java
private boolean hasEqualObject(List<ObjectNode> originalObjects, ObjectNode complexNode)
  {
    return originalObjects.parallelStream().anyMatch(originalNode -> {
      return originalNode.equals((o1, o2) -> {
        if (o1.isNumber() && o2.isNumber())
        {
          return o1.longValue() == o2.longValue() ? 0 : 1;
        }
        else
        {
          return o1.equals(o2) ? 0 : 1;
        }
      }, complexNode);
    });
  }
  ```
  
Did not see the method before. But I think it would be a good thing to support this by default. It shouldn't have any bad results. As long as the int-value is cast to long for the comparison it will never result in a bad ending. So it should be a good thing to add it natively to jackson :-)

**Reply by cowtowncoder** [reply]
Glad you were able to make it work. I appreciate your feedback; we are always open to improvements.
